velox/exec/tests/MultiFragmentTest.cpp

```diff
@@ -787,3 +788,147 @@ TEST_F(MultiFragmentTest, earlyCompletionMerge) {
     ASSERT_TRUE(waitForTaskCompletion(task.get())) << task->taskId();
   }
 }
+
+class CustomSlowNode : public core::PlanNode {
+ public:
+  CustomSlowNode(const core::PlanNodeId& id, core::PlanNodePtr source)
+      : PlanNode(id), sources_{std::move(source)} {}
+
+  const RowTypePtr& outputType() const override {
+    return sources_[0]->outputType();
+  }
+
+  const std::vector<core::PlanNodePtr>& sources() const override {
+    return sources_;
+  }
+
+  std::string_view name() const override {
+    return "custom slow";
+  }
+
+ private:
+  void addDetails(std::stringstream& /* stream */) const override {}
+
+  std::vector<core::PlanNodePtr> sources_;
+};
+
+class CustomSlowOperator : public Operator {
+ public:
+  CustomSlowOperator(
+      int32_t operatorId,
+      DriverCtx* driverCtx,
+      std::shared_ptr<const CustomSlowNode> slowNode)
+      : Operator(
+            driverCtx,
+            slowNode->outputType(),
+            operatorId,
+            slowNode->id(),
+            "CustomSlowOperator") {}
+
+  void addInput(RowVectorPtr input) override {
+    inputs_.emplace_back(std::move(input));
+  }
+
+  bool needsInput() const override {
+    return inputs_.empty();
+  }
+
+  RowVectorPtr getOutput() override {
+    if (inputs_.empty() || consumedOneBatch) {
+      return nullptr;
+    }
+    auto result = inputs_.back();
+    inputs_.pop_back();
+    consumedOneBatch = true;
+    return result;
+  }
+
+  void noMoreInput() override {
+    Operator::noMoreInput();
+  }
+
+  BlockingReason isBlocked(ContinueFuture* future) override {
+    return BlockingReason::kNotBlocked;
+  }
+
+  bool isFinished() override {
+    return false;
+  }
+
+ private:
+  std::vector<RowVectorPtr> inputs_;
+  bool consumedOneBatch{false};
+};
+
+class CustomSlowOperatorTranslator : public Operator::PlanNodeTranslator {
+  std::unique_ptr<Operator>
+  toOperator(DriverCtx* ctx, int32_t id, const core::PlanNodePtr& node) {
+    if (auto slowNode = std::dynamic_pointer_cast<const CustomSlowNode>(node)) {
+      return std::make_unique<CustomSlowOperator>(id, ctx, slowNode);
+    }
+    return nullptr;
+  }
+};
+
+TEST_F(MultiFragmentTest, exchangeDestruction) {
+  // This unit test tests the proper destruction of ExchangeClient upon
+  // task destruction.
+  OperatorTestBase::SetUp();
+  Operator::registerOperator(std::make_unique<CustomSlowOperatorTranslator>());
+  ExchangeSource::factories().clear();
```

Comment:
This will make it not possible to run tests in parallel. Consider using a different prefix for the custom exchange.

Reply:
Hi @mbasmanova could you explain more here? I am not sure if I understand it.

Reply:
When multiple tests are running in parallel, tests that rely on default factory may fail either because the factory has been unregistered or because the new factory registered in its place has a different behavior.